numpy/core/src/multiarray/multiarraymodule.c

```diff
@@ -3694,6 +3693,7 @@ _vec_string_with_args(PyArrayObject* char_array, PyArray_Descr* type,
         if (item_result == NULL) {
             goto err;
         }
+        Py_DECREF(args_tuple);
```

Comment:
Actually the `Py_XDECREF(args_tuple);` on line 3718 catches the error conditions, and moving the Py_DECREF up will, when there is an error, decref a `NULL` value. This is why I didn't move the declaration of arg_tuple into the loop - it would break the error handling